Fill.py

```diff
@@ -129,18 +142,35 @@ def distribute_items_restrictive(world: MultiWorld, fill_locations=None):
         else:
             restitempool.append(item)
 
-    world.random.shuffle(fill_locations)
-    call_all(world, "fill_hook", progitempool, nonexcludeditempool, localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+    call_all(world, "fill_hook", progitempool, nonexcludeditempool,
+             localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+
+    locationDeficit = len(progitempool) - len(prioritizedlocations)
+    if locationDeficit > 0:
+        if locationDeficit > len(defaultlocations):
+            raise FillError(
+                f'Not enough locations for progress items. There are {len(progitempool)} progression items with {len(prioritizedlocations)} priority locations and {len(defaultlocations)} default locations')
+        for i in range(0, locationDeficit):
+            prioritizedlocations.append(defaultlocations.pop())
+
+    fill_restrictive(world, world.state, prioritizedlocations, progitempool)
+    if prioritizedlocations:
+        defaultlocations = prioritizedlocations + defaultlocations
 
-    fill_restrictive(world, world.state, fill_locations, progitempool)
+    if progitempool:
+        fill_restrictive(world, world.state, defaultlocations, progitempool)
+
+    defaultlocations = defaultlocations + backfilllocations
```

Comment:
If you add `backfilllocations` to `defaultlocations` now, you will fill the excluded locations with never-excluded items. Moving this after the never-exclude fill fixes that.

Reply:
I added a tests and moved it. Since we don't add the excluded locations to the pool until after we have placed the non-excluded items, should we change the non-excluded items to use fast fill?

Reply:
where possible fast fill should be used

Reply:
Changed to fast fill for non excluded items